File: sdk/gsusb-cpp/examples/04_custom_timing.cpp
#include "demo_common.hpp"

#include <iostream>

int main(int argc, char** argv) {
    try {
        const auto options = demo::parse_common_options(argc, argv, "Single-channel custom timing CAN FD example");
        const auto [nominal, data] = demo::get_demo_custom_timings();
        auto tx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.tx_channel, true, nominal, data));
        auto rx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.rx_channel, true, nominal, data));

        gsusb::CanMessage message;
        message.arbitration_id = 0x7F0;
        message.is_fd = true;
        message.bitrate_switch = true;
        message.data = demo::make_payload({0x11}, 12, 0x00);

        demo::print_message("Sending", message);
        tx_bus->send(message);
        auto received = rx_bus->recv(std::chrono::seconds(2));
        if (!received.has_value()) {
            std::cerr << "No custom timing frame received\n";
            return 1;
        }
        demo::print_message("Received", received.value());
        tx_bus->shutdown();
        rx_bus->shutdown();
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << "\n";
        return 1;
    }
}

File: sdk/gsusb-csharp/examples/GsUsb.Examples/Examples/04_custom_timing.cs
using GsUsb;

namespace GsUsb.Examples;

internal static class Example04CustomTiming
{
    public static int Run(string[] args)
    {
        var options = CommonOptions.Parse(args, "Single-channel custom timing CAN FD example");
        var (nominal, data) = DemoCommon.GetDemoCustomTimings();
        using var txBus = new GsUsbBus(DemoCommon.MakeBusOptions(options, options.TxChannel, fd: true, nominal, data));
        using var rxBus = new GsUsbBus(DemoCommon.MakeBusOptions(options, options.RxChannel, fd: true, nominal, data));

        var message = new CanMessage
        {
            ArbitrationId = 0x7F0,
            IsFd = true,
            BitrateSwitch = true,
            Data = DemoCommon.MakePayload([0x11], 12, 0x00),
        };

        DemoCommon.PrintMessage("Sending", message);
        txBus.Send(message);
        var received = rxBus.Receive(TimeSpan.FromSeconds(2));
        if (received is null)
        {
            Console.Error.WriteLine("No custom timing frame received");
            return 1;
        }

        DemoCommon.PrintMessage("Received", received);
        return 0;
    }
}